### PR TITLE
Add decryption for iOS 10.2 Manifest.DB (fixes #14)

### LINKED ISSUE #14: sqlite3.DatabaseError: file is encrypted or is not a database
Following on from conversation in #13 and #10, I wasn't sure if this was the same or different problem so posting as a separate issue:

```
% python ./backup_tool.py /Users/shoffman/Library/Application\ Support/MobileSync/Backup/6460815e649d545617c60b38ff4efe459c32c935/
Device Name : sPhone6
Display Name : sPhone6
Last Backup Date : 2016-12-28 01:23:50
IMEI : XXXXXXXXXXX
Serial Number : XXXXXXXXXX
Product Type : iPhone7,1
Product Version : 10.2
iTunes Version : 12.5.4.42
Extract backup to /Users/shoffman/Library/Application Support/MobileSync/Backup/6460815e649d545617c60b38ff4efe459c32c935_extract ? (y/n)
y
Backup is encrypted
Enter backup password :
XXXXXXXXXXXXXXXXXXXXXXXXXX
Traceback (most recent call last):
  File "./backup_tool.py", line 104, in <module>
    main()
  File "./backup_tool.py", line 101, in main
    extract_backup(backup_path, output_path)
  File "./backup_tool.py", line 54, in extract_backup
    manifset_db = ManifestDB(backup_path)
  File "/Users/shoffman/workspace/iphone-dataprotection/python_scripts/backups/backup10.py", line 94, in __init__
    for record in cursor.execute("SELECT fileID, domain, relativePath, flags, file FROM Files"):
sqlite3.DatabaseError: file is encrypted or is not a database
```

This was using commit 241e52f.

Happy to help debug if somebody more familiar with the codebase can assist...

### PR DESCRIPTION
**Pull request by mbiette**
Hello,

I implemented the method from this comment to make iOS 10.2 backups work.
horrorho/InflatableDonkey#41 (comment)

And used it to decrypt the Manifest.db database.

Best regards,

Maxime

### CODE REVIEW

**Comment by seiterseiter** [comment]
How long is the "generating the password key for 10.2+" actually meant to take?

**Comment by BietteMaxime** [comment]
Because of the high iteration count it was taking 5 minutes in my tests to generate the password.

File: python_scripts/backups/backup10.py [hunk 1]
```diff
@@ -80,12 +80,20 @@ def is_directory(self):
...
         self.backup_path = path
         self.keybag = None
 
-        conn = sqlite3.connect(os.path.join(path,'Manifest.db'))
+        mdb_path = os.path.join(path,'Manifest.db')
+
+        #If a key is provided, try to decrypt the DB
+        if key:
+            mdb_path_encrypted = mdb_path
+            mdb_path = os.path.join(path,'Manifest.db-decypted')
```
Comment (uberhitzt) [inline python_scripts/backups/backup10.py]: typo in filename

Comment (BietteMaxime) [inline python_scripts/backups/backup10.py]: Made a change.
Thank you

**Review by uberhitzt — APPROVED** [review]

**Review by uberhitzt — APPROVED** [review]
Thanks! This works on my backups from 10.2 & 10.3

**Comment by dinosec** [comment]
Thanks @BietteMaxime for the contribution for iOS 10.2!

**Comment by JakeM37** [comment]
Hello,

I'm new at GitHub and new at using python.

I really need your help to get this Skript running on my backup.
I have already tried it but getting always this issue 

> $

 python /Users/peterkoch/Desktop/iphone-dataprotection-master/python_scripts/backup_tool.py 
Traceback (most recent call last):
  File "/Users/peterkoch/Desktop/iphone-dataprotection-master/python_scripts/backup_tool.py", line 2, in <module>
    from backups.backup3 import decrypt_backup3
  File "/Users/peterkoch/Desktop/iphone-dataprotection-master/python_scripts/backups/backup3.py", line 2, in <module>
    from crypto.aes import AESdecryptCBC
  File "/Users/peterkoch/Desktop/iphone-dataprotection-master/python_scripts/crypto/aes.py", line 1, in <module>
    from Crypto.Cipher import AES
ImportError: No module named 

> Crypto.Cipher

maybe you can help me?

I'm using a macOS 10.12.6 Sierra
Thanks guys!

**Comment by marcas91** [comment]
You're missing the crypto.cipher module. Install it with pip:

$ sudo easy_install pip
$ easy_install pycrypto

**Comment by JakeM37** [comment]
Thanks @marcas91 it worked.

But know I've a new issue 

> Peters-iMac:python_scripts peterkoch$ python3 backup_tool.py
>   File "backup_tool.py", line 17
>     print "Manifest.plist not found"
>                                    ^
> SyntaxError: Missing parentheses in call to 'print'

It would be grateful if somebody help me to get starting this script.

I apologize for my bad english and also for being so noobie

**Comment by marcas91** [comment]
You're using Python 3. Uninstall it, and install Python 2. There were many syntax changes from 2 to 3.

**Comment by JakeM37** [comment]
People I desperately slow can someone help me to get the script to run or to decrypt the backup. With python2 it did not work either

**Comment by lgrimani** [comment]
Does the ManifestKey/BackupKeybag change completely with each incremental differential backup?  I tested this a few times and sometimes it doesn't change at all and sometimes some of the contents of the Keybag change but its not clear to me how and when.

Once Manifest.DB is decrypted, does each encrypted file have its own unique key that needs to be generated from Manifest.DB?

Finally, when entries are removed from the Files table, are they removed completely from the SQLite database or is it just hidden by a delete flag?

I am trying to decrypt encrypted files from one backup ago where a newer Manifest.DB/.plist has overwritten the old ones.

**Pull request merged**